### PR TITLE
feedbacks: async call updation, code optimized for rendering books

### PR DESCRIPTION
**Pull request by PrathamGupta-Compro**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,17 +4,19 @@ const priceMinFilter = document.querySelector('#input-search-price-min');
...
-const searchBtn = document.querySelector('#btn-search');
-const tableBody = document.querySelector('#table-similar-books tbody');
-const tableth = document.querySelector('table th');
+const similarTableBody = document.querySelector('#table-similar-books tbody');
+const allBooksTableBody = document.querySelector('#table-all-books tbody');
 const similarBookContainer = document.querySelector('.similar-book-container');
 const defaultImage = './product-not-found.png';
 const similarBookCount = document.querySelector('#similar-book-count');
+// const upArrowSort = document.querySelector('.fa-arrow-up-short-wide');
+// const downArrowSort = document.querySelector('.fa-arrow-down-short-wide');
```
Comment (shubham-shukla-compro) [inline index.js]: remove the commented code. if not needed

File: index.js [hunk 2]
```diff
@@ -83,16 +39,17 @@ const renderPagination = (totalPages) => {
...
+        renderAllBooks();
+      }
     });
     return li;
   };
 
   paginationElement.appendChild(createPageItem(currentPage - 1, 'Previous'));
 
-  for (let i = 1; i <= totalPages; i++) {
+  for (let i = 1; i <= totalPages; i += 1) {
```
Comment (shubham-shukla-compro) [inline index.js:71]: why this change? any difference in `i++` and `i += 1`

File: index.js [hunk 3]
```diff
@@ -141,262 +95,224 @@ const filterData = (val, filterType, currentData) => {
...
   console.log('Filtered Examined Books :- ');
   console.log(currentData);
   return currentData;
 };
 
-//Here We Want to render Similar Books, firstly on the basis of the same genre and then on the basis of ascending order of price.
-//RULE: We want to display similar books with same genre and similarPrice Range in sorted Order, then the rest.
+// Here We Want to render Similar Books, firstly on the basis of the same genre and then on the basis of ascending order of price.
+// RULE: We want to display similar books with same genre and similarPrice Range in sorted Order, then the rest.
 const renderSimilarBooks = (currentBook) => {
```
Comment (shubham-shukla-compro) [inline index.js:206]: This method is too long and seems unoptimized, we can rewrite the method like below. Please verify
```
const renderSimilarBooks = (currentBook) => {
  const similarBooksMap = new Map();

  data.forEach((book) => {
    if (book.bookId !== currentBook.bookId) {
      const matchesGenre = book.genre.toLowerCase() === currentBook.genre.toLowerCase();
      const matchesPrice = book.price >= currentBook.price * 0.9 && book.price <= currentBook.price * 1.1;

      if (matchesGenre && matchesPrice) {
        similarBooksMap.set(book.bookId, book);
      } else if (matchesGenre || matchesPrice) {
        if (!similarBooksMap.has(book.bookId)) {
          similarBooksMap.set(book.bookId, book);
        }
      }
    }
  });

  const finalSimilarBookList = Array.from(similarBooksMap.values());

  console.log('All Similar Books :- ');
  console.log(finalSimilarBookList);
  
  similarTableBody.innerHTML = ''; // Clear old rows to avoid duplicates

  finalSimilarBookList.slice(0, 10).forEach((book) => {
    const coverImage = book.coverImage || defaultImage;
    const rowEntry = document.createElement('tr');
    rowEntry.innerHTML = `
      <td><img src="${coverImage}" alt="${book.bookName}" style="width:50px;height:50px; object-fit:contain;"></td>
      <td>${book.bookName}</td>
      <td>${book.genre}</td>
      <td>$ ${book.price}</td>
      <td>${book.author}</td>
      <td>${book.publicationYear}</td>`;
    similarTableBody.appendChild(rowEntry);
  });

  similarBookCount.innerHTML = `Showing <b>${Math.min(10, finalSimilarBookList.length)}</b> of <b>${finalSimilarBookList.length}</b> Similar Book(s)`;
  renderAllBooks();
};

```

File: index.js [hunk 4]
```diff
@@ -405,34 +321,96 @@ function loadPageState() {
...
+      data.sort((a, b) => (isAsc ? a.price - b.price : b.price - a.price));
+    } else if (valType === 'year') {
+      data.sort((a, b) => (isAsc ? a.publicationYear - b.publicationYear : b.publicationYear - a.publicationYear));
+    }
+    isAsc = !isAsc;
+    renderAllBooks();
+  }
+};
+
+const handleSearch = () => {
```
Comment (shubham-shukla-compro) [inline index.js:508]: This code for **handleSearch** and **filterData** looks unoptimized and non-readable, can you try optimizing like below and test.
```
const filterData = (val, filterType, currentData) => {
  const lowerCaseVal = val.toLowerCase();
  switch (filterType) {
    case 'bookName':
      return currentData.filter((book) => book.bookName.toLowerCase().includes(lowerCaseVal));
    case 'genre':
      return currentData.filter((book) => book.genre.toLowerCase().includes(lowerCaseVal));
    case 'priceMin':
      return currentData.filter((book) => book.price >= val);
    case 'priceMax':
      return currentData.filter((book) => book.price <= val);
    case 'author':
      return currentData.filter((book) => book.author.toLowerCase().includes(lowerCaseVal));
    case 'publicationYear':
      return currentData.filter((book) => book.publicationYear.toString() === val);
    default:
      return currentData;
  }
};

const handleSearch = () => {
  let filteredData = data;
  const filters = [
    { value: bookNameFilter.value, type: 'bookName' },
    { value: bookGenreFilter.value, type: 'genre' },
    { value: priceMinFilter.value, type: 'priceMin' },
    { value: priceMaxFilter.value, type: 'priceMax' },
    { value: bookAuthorFilter.value, type: 'author' },
    { value: yearFilter.value, type: 'publicationYear' }
  ];

  filters.forEach(filter => {
    if (filter.value) {
      filteredData = filterData(filter.value, filter.type, filteredData);
    }
  });

  if (filters.some(filter => filter.value)) {
    filteredData.sort((a, b) => a.price - b.price);
    similarBookContainer.style.display = 'block';
    renderSearchedBookDetails(filteredData);
  } else {
    renderSearchedBookDetails([]);
  }

  savePageState();
};
```

File: index.js (hunk 1 above)
Comment (PrathamGupta-Compro) [reply index.js]: Done

File: index.js (hunk 2 above)
Comment (PrathamGupta-Compro) [reply index.js:71]: ESLint error: changed to i++

File: index.js (hunk 3 above)
Comment (PrathamGupta-Compro) [reply index.js:206]: Done. I have made all the required changes but, the code you've given for rendering similar books seems inconsistent and it doesn't display the data in required order. So I have replaced the filtering logic with my code and the rendering logic with the code you've provided.

File: index.js (hunk 4 above)
Comment (PrathamGupta-Compro) [reply index.js:508]: Done

**Review by shubham-shukla-compro — APPROVED** [review]
